### PR TITLE
Reader-writer lock and capsule API with shared

### PR DESCRIPTION
**Pull request by ageorg29**
Added a `rwlock` to the capsule API to support many-readers with `shared` access, and one-writer with `uncontended` access.
 
`create_with_rwlock` creates a capsule with a reader-writer lock. 

The reader-writer lock module has three functions: `with_write_lock` works like `with_lock` using a mutex, `with_read_lock` grants access to a new `ReaderPassword.t` type, and `destroy` destroys a reader-writer lock by setting its poison bit to 1.

A ReaderPassword.t is required to call `Data.map_shared` and `Data.extract_shared`. It is safe to turn a `Password.t` into a `ReaderPassword.t`.

`Data.map_shared` and `Data.extract_shared` both grant `shared` access into a capsule. Note that the internal type of the `Data.t` input must cross portability. This is to prevent indirect write access via closures that might enclose `uncontended` access to the capsule's data.

### CODE REVIEW

**Review by stedolan — COMMENTED** [review]
The C code here looks good to me (in particular, it's doing the right thing with respect to blocking sections and GC), but I have a few questions about how it's to be exposed safely to OCaml.

File: ocaml/otherlibs/stdlib_alpha/capsule.ml [hunk 1]
```diff
@@ -80,9 +105,67 @@ module Mutex = struct
...
+    @@ portable
+    = fun t f ->
+      Rw.lock_read t.rwlock;
+      match t.poisoned with
+      | true -> Rw.unlock t.rwlock; reraise Poisoned
+      | false ->
+        match f (ReaderPassword.unsafe_mk()) with
+        | x -> Rw.unlock t.rwlock; x
+        | exception exn ->
+          t.poisoned <- true;
```
[the remark this answers is not in the document]
Comment (ageorg29) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:146]: Yes indeed! That's bad. I can make the poison bit into an Atomic

[the remark this answers is not in the document]
Comment (lpw25) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:146]: I think we can probably not poison on exceptions from a reader, since they shouldn't be leaving the protected data in a bad state.

File: ocaml/otherlibs/stdlib_alpha/capsule.ml [hunk 2]
```diff
@@ -80,9 +105,67 @@ module Mutex = struct
...
 
+module Rwlock = struct
+
+  type 'k t : value mod portable uncontended = { rwlock : Rw.t; mutable poisoned : bool }
+
+  type packed : value mod portable uncontended = P : 'k t -> packed
+
+  exception Poisoned
+
+  let with_write_lock :
```
[the remark this answers is not in the document]
Comment (stedolan) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:116]: After some further discussion / googling: This is fine.

I referenced the wrong version of POSIX (actually even older than POSIX: my link is to the SUSv2). The current version specifies that write locking a rwlock locked by the same thread is either a EDEADLK or a deadlock, not general UB (and therefore "safe").

Here's the discussion on clarifying this point in POSIX: https://www.austingroupbugs.net/view.php?id=720

(At some point it may be worth testing actual implementations on this point - in particular, the internet suggests that macos may do the wrong thing here)

File: ocaml/otherlibs/stdlib_alpha/capsule.ml [hunk 3]
```diff
@@ -80,9 +105,67 @@ module Mutex = struct
       Password.unsafe_mk ()
 end
 
+module Rwlock = struct
+
+  type 'k t : value mod portable uncontended = { rwlock : Rw.t; mutable poisoned : bool }
```
[the remark this answers is not in the document]
Comment (stedolan) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:110]: Explanation: it's the `-allow-illegal-crossing` in `ocaml/otherlibs/stdlib_alpha/dune` that allows this.

File: ocaml/otherlibs/stdlib_alpha/rwlock.c [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+#ifdef CAML_RUNTIME_5
+#include "../../runtime/sync_posix.h"
+
+/* Rwlock ops */
+static void caml_rwlock_finalize(value wrapper)
+{
+  sync_rwlock_destroy(Rwlock_val(wrapper));
+}
+
+static int caml_rwlock_compare(value wrapper1, value wrapper2)
```
[the remark this answers is not in the document]
Comment (ageorg29) [reply ocaml/otherlibs/stdlib_alpha/rwlock.c:21]: I added the nontrivial implementations to mirror the Mutex implementation, which has similar `compare` and `hash` functions, I can't think of a particular reason to use nontrivial implementations

File: ocaml/otherlibs/stdlib_alpha/capsule.ml [hunk 5]
```diff
@@ -80,9 +105,67 @@ module Mutex = struct
...
+
+  let with_write_lock :
+    'k t
+    -> ('k Password.t @ local -> 'a) @ local
+    -> 'a
+    @@ portable
+    = fun t f ->
+      Rw.lock_write t.rwlock;
+      match t.poisoned with
+      | true -> Rw.unlock t.rwlock; reraise Poisoned
```
[the remark this answers is not in the document]
Comment (ageorg29) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:124]: The `Mutex.with_lock` implementation makes a similar reraise if the poison flag is true, should both be changed, or just this one on line 124?

File: ocaml/otherlibs/stdlib_alpha/rwlock.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (stedolan) [reply ocaml/otherlibs/stdlib_alpha/rwlock.c:21]: Makes sense!

File: ocaml/otherlibs/stdlib_alpha/capsule.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (stedolan) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:124]: Yeah, lines 124 and 141 should both be changed.

For context: When an exception is raised, the runtime collects a backtrace for the portion of stack being discarded, between the raise and the catch. If the handler for that exception uses `reraise` to continue propagating the exception, then the next piece of backtrace is _appended_ to the existing backtrace. This mechanism allows the backtrace for e.g. `with_write_lock lk (fun _ -> failwith "blah")` to point at the `failwith`, even though the exception was most recently raised by the handler logic inside `with_write_lock`.

This mechanism only makes sense when the runtime's global backtrace buffer contains something related to the exception you're about to reraise. In the case where `t.poisoned` is true, the backtrace buffer contains whatever happens to be the most recent backtrace is whatever code most recently raised. Appending to that backtrace is not going to give meaningful results.

In other words, `reraise` only really makes sense inside the exception-handling arm of a `try`, which describes some but not all of the uses of `reraise` here.

(If you really wanted to, you could use reraise here by capturing the backtrace of the point that failed when setting `poisoned <- true`, and restoring it using `Printexc.raise_with_backtrace`, to guarantee you're appending to the correct backtrace. I don't recommend doing this, but that's what correct uses of reraise outside of exception handlers have to look like)

[the remark this answers is not in the document]
Comment (ageorg29) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:124]: Interesting! And yes the explanation makes sense. In total I think there are 5 uses of `reraise` outside an exception-handling arm: lines 86, 101, 124 and 144 and 155. If I understand correctly, we could replace all those with a `raise`? 

Out of curiosity, what happens if an exception is re-raised via `reraise` in one thread (by catching a exception being thrown by some call-back function f), after which another thread then raises a `Poisoned` exception via `raise`, does each thread have its own backtrace?

[the remark this answers is not in the document]
Comment (stedolan) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:124]: > If I understand correctly, we could replace all those with a raise?

Yes - I hadn't spotted those other ones!

> does each thread have its own backtrace?

Yes, in order to handle the situation you describe. (If you're curious, grep for `backtrace_buffer` in `ocaml/runtime` and `ocaml/otherlibs/systhreads`)

[the remark this answers is not in the document]
Comment (lpw25) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:124]: I believe @stedolan has forgotten that on our branch `raise` is defined as `%reraise` for confusing semi-historical reasons. I think the right fix here is actually to rename the `external reraise ...` in this file to `external raise ...`

[the remark this answers is not in the document]
Comment (lpw25) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:124]: I'm going to resolve this, since the confusing name for `raise` is unrelated to this PR.

**Comment by lpw25** [comment]
This needs a rebase to fix some conflicts. Then I think it is ready to go.

File: ocaml/otherlibs/stdlib_alpha/capsule.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (stedolan) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:124]: Nope, haven't forgotten that, and I still this this is a problem. Since we define `raise` as `reraise`, our backtraces are generally broken. When we fix this, the backtraces elsewhere will improve, but backtraces arising from poisoned mutexes will still be wrong due to this redefinition.

(I agree it's an instance of a thing that is generally broken elsewhere, but since it's being explicitly redefined here we may as well get it right this time)

[the remark this answers is not in the document]
Comment (lpw25) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:124]: Hmm, I guess. `raise` is only being re-defined here because the main one isn't portable yet. As soon as it is I think we should call it normally here rather than redefine it. In which case its behaviour will go straight back to `%reraise`, so I don't think I see the benefit of using `%raise` here temporarily.

[the remark this answers is not in the document]
Comment (lpw25) [reply ocaml/otherlibs/stdlib_alpha/capsule.ml:124]: If you feel strongly about it maybe change the existing definition in a separate PR? It seems unrelated to this PR.

**Review by stedolan — APPROVED** [review]
